wandb/filesync/step_prepare.py

```diff
@@ -1,5 +1,7 @@
 """Batching file prepare requests to our API."""
```

Comment:
general question, does it make sense to have 2 classes one for sync and one for async (maybe for the Artifacts, not sure about other). Not sure yet if it is a good idea, just something that came to mind, while reading the PR

Reply:
Very reasonable question! Yeah, having two classes _would_ feel cleaner to me, and I was originally aiming for that. But the async implementation sometimes needs to fall back to the sync implementation:
1. the async impl [doesn't (yet?) support Azure file-uploads](https://github.com/wandb/wandb/blob/ad7711f0abd7b192d2856163a4de97cf2dadce72/wandb/sdk/internal/internal_api.py#L2064-L2066), so it falls back to the sync impl
2. if the async upload fails, [it falls back to the sync impl](https://github.com/wandb/wandb/blob/ad7711f0abd7b192d2856163a4de97cf2dadce72/wandb/filesync/step_upload.py#L304-L311) (tell me if you think I'm just being retry-happy here; maybe it should simply fail)

This fallback behavior would be _possible_ to do with separate `(Sync|Async)StepUpload` classes, but it would be... _differently_ complicated. The AsyncStepUpload would need to have a handle on the SyncStepUpload; sometimes (in case (2) above) "falling back" would consist of sending the SyncStepUpload a `RequestUpload` message, but sometimes (in (1)) it would require, not direct interaction with the SyncStepUpload, but rather, scheduling a task in its thread-pool.

(Also, there would be a fair bit of shared behavior between those two classes, which I'd want to refactor out into a common superclass or something; and I'm trying to _avoid_ modifying the existing code-paths too much in this PR.)

Reply:
Yeah, that totally makes sense to me and i like that we have a fallback approach and not just fail. We can discuss in more details, but i would be interested to understand how the async fails, to make sure we don't find ourselves in a weird state.